python/rapidsmpf/rapidsmpf/integrations/core.py

```diff
@@ -0,0 +1,469 @@
+        shuffler: Shuffler,
+        options: Any,
+        *other: Any,
+    ) -> None:
+        """
+        Add a partition to a RapidsMPF Shuffler.
+
+        Parameters
+        ----------
+        df
+            DataFrame partition to add to a RapidsMPF shuffler.
+        partition_id
+            The input partition id of ``df``.
+        partition_count
+            Number of output partitions for the current shuffle.
+        shuffler
+            The RapidsMPF Shuffler object to extract from.
+        options
+            Additional options.
+        *other
+            Other data needed for partitioning. For example,
+            this may be boundary values needed for sorting.
+        """
+
+    @staticmethod
+    def extract_partition(
+        partition_id: int,
+        shuffler: Shuffler,
+        options: Any,
+    ) -> DataFrameT:
+        """
+        Extract a DataFrame partition from a RapidsMPF Shuffler.
+
+        Parameters
+        ----------
+        partition_id
+            Partition id to extract.
+        shuffler
+            The RapidsMPF Shuffler object to extract from.
+        options
+            Additional options.
+
+        Returns
+        -------
+        A shuffled DataFrame partition.
+        """
+
+
+def get_shuffler(
+    get_context: Callable[..., WorkerContext],
+    shuffle_id: int,
+    *,
+    partition_count: int | None = None,
+    worker: Any = None,
+) -> Shuffler:
+    """
+    Return the appropriate :class:`Shuffler` object.
+
+    Parameters
+    ----------
+    get_context
+        Callable function to fetch the worker context.
+    shuffle_id
+        Unique ID for the shuffle operation.
+    partition_count
+        Output partition count for the shuffle operation.
+    worker
+        The current worker.
+
+    Returns
+    -------
+    The active RapidsMPF :class:`Shuffler` object associated with
+    the specified ``shuffle_id``, ``partition_count`` and
+    ``worker``.
+
+    Notes
+    -----
+    Whenever a new :class:`Shuffler` object is created, it is
+    saved as ``WorkerContext.shufflers[shuffle_id]``.
+    """
+    ctx = get_context(worker)
+    with ctx.lock:
+        if shuffle_id not in ctx.shufflers:
+            if partition_count is None:
+                raise ValueError(
+                    "Need partition_count to create new shuffler."
+                    f" shuffle_id: {shuffle_id}\n"
+                    f" Shufflers: {ctx.shufflers}"
+                )
+            assert ctx.br is not None
+            assert ctx.comm is not None
+            assert ctx.progress_thread is not None
+            ctx.shufflers[shuffle_id] = Shuffler(
+                ctx.comm,
+                ctx.progress_thread,
+                op_id=shuffle_id,
+                total_num_partitions=partition_count,
+                stream=DEFAULT_STREAM,
+                br=ctx.br,
+                statistics=ctx.statistics,
+            )
+    return ctx.shufflers[shuffle_id]
```

Comment:
```suggestion
        return ctx.shufflers[shuffle_id]
```

This return needs to happen with the lock held I think.

Reply:
Good catch, we should hold on to the lock. 

PS: the current version without the lock works in CPython because `dict[x]` is atomic.